samples/vulkan/ocean/ocean.hpp

```diff
@@ -0,0 +1,327 @@
+        IOPT_NORMAL_MAP,
+        IOPT_COUNT
+    };
+
+    struct TextureOCL
+    {
+        std::vector<VkImage> images;
+        std::vector<VkDeviceMemory> image_memories;
+        std::vector<VkImageView> image_views;
+    };
+
+    // vulkan-opencl interop resources
+    std::array<TextureOCL, IOPT_COUNT> texture_images;
+
+    // Ocean grid vertices and related buffers
+    std::vector<Vertex> ocean_grid_vertices;
+    std::vector<VkBuffer> vertex_buffers;
+    std::vector<VkDeviceMemory> vertex_buffer_memories;
+
+    std::vector<std::uint32_t> ocean_grid_indices;
+    // separate index buffer for [ocean_grid_size] triangle strips
+    struct IndexBuffer
+    {
+        std::vector<VkBuffer> buffers;
+        std::vector<VkDeviceMemory> buffer_memories;
+    };
+    std::vector<IndexBuffer> index_buffers;
+
+    std::array<VkSampler, IOPT_COUNT> texture_sampler;
+
+    VkDescriptorPool descriptor_pool;
+    std::vector<VkDescriptorSet> descriptor_sets;
+
+    std::vector<VkCommandBuffer> command_buffers;
+
+    std::vector<VkSemaphore> image_available_semaphores;
+    std::vector<VkSemaphore> render_finished_semaphores;
+    std::vector<VkFence> in_flight_fences;
+    std::vector<VkFence> images_in_flight;
+    size_t current_frame = 0;
+
+#ifdef _WIN32
+    PFN_vkGetMemoryWin32HandleKHR vkGetMemoryWin32HandleKHR = NULL;
+#elif defined(__linux__)
+    PFN_vkGetMemoryFdKHR vkGetMemoryFdKHR = NULL;
+#endif
+
+    std::vector<VkBuffer> uniform_buffers;
+    std::vector<VkDeviceMemory> uniform_buffers_memory;
+
+    // more OpenCL resources
+    cl_external_memory_handle_type_khr external_mem_type = 0;
+
+    cl::Context context;
+    cl::CommandQueue command_queue;
+
+    // generates twiddle factors kernel
+    cl::Kernel twiddle_kernel;
+
+    // initial spectrum kernel
+    cl::Kernel init_spectrum_kernel;
+
+    // Fourier components image kernel
+    cl::Kernel time_spectrum_kernel;
+
+    // FFT kernel
+    cl::Kernel fft_kernel;
+
+    // inversion kernel
+    cl::Kernel inversion_kernel;
+
+    // building normals kernel
+    cl::Kernel normals_kernel;
+
+    // FFT intermediate computation storages without vulkan iteroperability
+    std::unique_ptr<cl::Image2D> dxyz_coef_mem[3];
+    std::unique_ptr<cl::Image2D> hkt_pong_mem;
+    std::unique_ptr<cl::Image2D> twiddle_factors_mem;
+    std::unique_ptr<cl::Image2D> h0k_mem;
+    std::unique_ptr<cl::Image2D> noise_mem;
+
+    size_t ocl_max_img2d_width;
+    cl_ulong ocl_max_alloc_size, ocl_mem_size;
+
+    // main opencl-vulkan iteroperability resources
+    // final computation result with displacements and normal map,
+    // needs to follow swap-chain scheme
+    std::array<std::vector<std::unique_ptr<cl::Image2D>>, IOPT_COUNT>
+        ocl_image_mems;
+
+    clEnqueueAcquireExternalMemObjectsKHR_fn
+        clEnqueueAcquireExternalMemObjectsKHR = NULL;
+    clEnqueueReleaseExternalMemObjectsKHR_fn
+        clEnqueueReleaseExternalMemObjectsKHR = NULL;
```

Comment:
Having these function pointers won't hurt, but this sample is currently calling into the C++ bindings (OpenCL-CLHPP) to acquire and release the external memory objects.  The C++ bindings query their own function pointers, so I'm pretty sure these function pointers are never used and can be removed.

Reply:
Indeed, thanks